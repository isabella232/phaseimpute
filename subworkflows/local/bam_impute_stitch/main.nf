include { GAWK                 } from '../../../modules/nf-core/gawk'
include { STITCH               } from '../../../modules/nf-core/stitch'
include { BCFTOOLS_INDEX       } from '../../../modules/nf-core/bcftools/index'

workflow BAM_IMPUTE_STITCH {

    take:
    ch_input        // channel:   [ [id], bam, bai, bamlist ]
    ch_posfile      // channel:   [ [panel, chr], legend ]
    ch_region       // channel:   [ [chr, region], region ]
    ch_fasta        // channel:   [ [genome], fa, fai ]

    main:

    ch_versions      = Channel.empty()
    // Run STITCH
    seed = params.seed

    // Value channels
    def input_empty         = [[]]
    def rdata_empty         = [[]]
    k_val_params            = params.k_val
    ngen_params             = params.ngen

    // Transform posfile to TSV with ','
    GAWK(ch_posfile, [])
    ch_versions = ch_versions.mix(GAWK.out.versions)

    // Get chromosomes of posfile
    ch_posfile = GAWK.out.output
        .map{metaPC, posfile -> [[chr: metaPC.chr], metaPC, posfile]}

    // Get chromosomes of fasta
    ch_chromosomes = ch_region
        .map{metaCR, region -> [[chr: metaCR.chr], metaCR.chr]}

    // Make final channel with parameters
    ch_parameters = ch_posfile
        .map { it + input_empty + rdata_empty}
        .join(ch_chromosomes)
        .map { it + k_val_params + ngen_params}
        .map { metaC, metaPC, posfile, input, rdata, chr, k_val, ngen ->
            [metaPC, posfile, input, rdata, chr, k_val, ngen]
        }

    ch_bam_params = ch_input // Add chr to meta map
        .combine(ch_parameters)
        .map{
            metaI, bam, bai, bamlist, metaPC, posfile, input, rdata, chr, k_val, ngen ->
            [
                metaI + [chr: metaPC.chr, panel:metaPC.id],
                bam, bai, bamlist, posfile, input, rdata, chr, k_val, ngen
            ]
        }

    STITCH( ch_bam_params, ch_fasta, seed )
    ch_versions = ch_versions.mix(STITCH.out.versions)

    // Index imputed annotated VCF
    BCFTOOLS_INDEX(STITCH.out.vcf)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX.out.versions)

    // Join VCFs and TBIs
    ch_vcf_tbi = STITCH.out.vcf
        .join(BCFTOOLS_INDEX.out.tbi)
        .map { metaI, vcf, tbi -> [ metaI + [tools: "stitch"], vcf, tbi ] }

    emit:
    vcf_tbi  = ch_vcf_tbi                        // channel:   [ [id, chr], vcf, tbi ]
    versions = ch_versions                       // channel:   [ versions.yml ]

}
