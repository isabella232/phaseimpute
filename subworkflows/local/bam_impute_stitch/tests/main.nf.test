nextflow_workflow {

    name "Test Subworkflow BAM_IMPUTE_STITCH"
    script "../main.nf"
    config "./nextflow.config"

    workflow "BAM_IMPUTE_STITCH"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_impute_stitch"
    tag "bam_impute_stitch"

    tag "stitch"
    tag "bcftools"
    tag "bcftools_index"

    test("Impute with stitch two individuals with bamlist") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                bamlist = Channel.of(
                    "NA12878.s.bam",
                    "NA19401.s.bam"
                ).collectFile(name: 'bamlist.txt', newLine: true)
                ch_samples = Channel.fromList([
                    [
                        [id: "allid"],
                        [
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true)
                        ],
                        [
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                            file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true)
                        ]
                    ],
                ])

                ch_posfile = Channel.of([
                    [id: "1000GP", chr: "chr22"],
                    file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                ])

                ch_region = Channel.of([
                    [chr: "chr22", region: "chr22:16570000-16610000"], "chr22:16570000-16610000"
                ])

                input[0] = ch_samples.combine(bamlist)
                input[1] = ch_posfile
                input[2] = ch_region
                input[3] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out,
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.summary },
                    workflow.out.vcf_tbi.collect{ path(it[1]).vcf.header.getGenotypeSamples().sort() }
                ).match() }
            )
        }
    }
}
