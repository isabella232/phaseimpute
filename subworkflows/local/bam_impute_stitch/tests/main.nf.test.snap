{
    "Impute with stitch two individuals with bamlist": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "allid",
                            "chr": "chr22",
                            "panel": "1000GP",
                            "tools": "stitch"
                        },
                        "stitch.chr22.vcf.gz:md5,bd5f0c37e08b601ba1cd9ffc8fd672fa",
                        "stitch.chr22.vcf.gz.tbi:md5,bc2655a04da76a431481799d3a4515ee"
                    ]
                ],
                "1": [
                    "versions.yml:md5,4330856a215c4986b0591d550a4d22bf",
                    "versions.yml:md5,8561197b98b9a1e1f8f1b01c1369da35",
                    "versions.yml:md5,c0772dccc2cc2a3c8f0b3fa5f02312bc"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "allid",
                            "chr": "chr22",
                            "panel": "1000GP",
                            "tools": "stitch"
                        },
                        "stitch.chr22.vcf.gz:md5,bd5f0c37e08b601ba1cd9ffc8fd672fa",
                        "stitch.chr22.vcf.gz.tbi:md5,bc2655a04da76a431481799d3a4515ee"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4330856a215c4986b0591d550a4d22bf",
                    "versions.yml:md5,8561197b98b9a1e1f8f1b01c1369da35",
                    "versions.yml:md5,c0772dccc2cc2a3c8f0b3fa5f02312bc"
                ]
            },
            [
                "VcfFile [chromosomes=[chr22], sampleCount=2, variantCount=903, phased=false, phasedAutodetect=false]"
            ],
            [
                [
                    "NA12878",
                    "NA19401"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-06T16:29:44.891707584"
    }
}